keps/sig-node/24-apparmor/README.md

```diff
@@ -0,0 +1,659 @@
+
+If built-in support for multiple runtime profiles is needed in the future, a new
+KEP will be created to cover its details.
+
+##### Localhost Profile
+
+This KEP proposes LocalhostProfile as the only source of user-defined
+profiles. User-defined profiles are essential for users to realize
+the full benefits out of AppArmor, allowing them to decrease their attack
+surface based on their own workloads.
+
+###### Updating localhost AppArmor profiles
+
+AppArmor profiles are applied at container creation time. The underlying
+container runtime only references already loaded profiles by its name.
+Therefore, updating the profiles content requires a manual reload (typically via
+`apparmor_parser`).
+
+Note that changing profiles is not recommended and may cause containers to fail
+on next restart, in the case of the new profile being more restrictive, invalid
+or the file no longer available on the host.
+
+Currently, users have no way to tell whether their physical profiles have been
+deleted or modified. This KEP proposes no changes to the existing functionality.
+
+The recommended approach for rolling out changes to AppArmor profiles is to
+always create _new profiles_ instead of updating existing ones. Create and
+deploy a new version of the existing Pod Template, changing the profile name to
+the newly created profile. Redeploy, once working delete the former Pod
+Template. This will avoid disruption on in-flight workloads.
+
+The current behavior lacks features to facilitate the maintenance of AppArmor
+profiles across the cluster. Two examples being: 1) the lack of profile
+synchronization across nodes and 2) how difficult it can be to identify that
+profiles have been changed on disk/memory, after pods started using it. However,
+Kubernetes managed profiles are out of scope for this KEP.
+Out of tree enhancements like the
+[security-profiles-operator](https://github.com/kubernetes-sigs/security-profiles-operator) can
+provide such enhanced functionality on top.
+
+###### Profiles managed by the cluster admins
+
+The current support relies on profiles being loaded on all cluster nodes
+where the pods using them may be scheduled. It is also the cluster admin's
+responsibility to ensure the profiles are correctly saved and synchronized
+across the all nodes. Existing mechanisms like node `labels` and `nodeSelectors`
+can be used to ensure that pods are scheduled on nodes supporting their desired
+profiles.
+
+#### Validation
+
+The following validations were applied to the AppArmor annotations on pods:
+- Pod annotations are immutable (cannot be added, modified, or removed on pod update)
+- Annotation value must have a `localhost/` prefix, or be one of: `""`, `runtime/default`, `unconfined`.
+
+The annotation validations will be carried over to the field API, and the following additional
+validations are proposed:
+1. Fields must match the corresponding annotations when both are present, except for ephemeral containers.
```

Comment:
Just checking my understanding-- When annotations are set, non-ephemeral containers fields must match the annotations, but ephemeral containers fields may be set to any desired values?

Reply:
Yes. This is due to the fact that the annotations are currently immutable, and that restriction wasn't relaxed when ephemeral containers were introduced. The fields will still be immutable, but when an ephemeral container is added, it may set an AppArmor profile on its security context. In this case, the annotations won't be synced.